### PR TITLE
[AMDGPU] Fix subtarget predicates for some V_MAD, V_FMA and V_DOT instructions.

### PR DESCRIPTION
**Pull request by kosarev**
Resolves AsmParser ambiguities, e.g., between V_FMA_MIXLO_F16_vi and V_FMA_MIXLO_F16_gfx11.

Part of <https://github.com/llvm/llvm-project/issues/69256>.

### CODE REVIEW

**Review by jayfoad — APPROVED** [review]
LGTM.

How do you find these problems? Once they are all fixed can you add some kind of assertion somewhere to stop them reoccuring?

**Review by Sisyph — APPROVED** [review]
LGTM

**Comment by kosarev** [comment]
> How do you find these problems? Once they are all fixed can you add some kind of assertion somewhere to stop them reoccuring?

I use the ambiguity checks in AsmMatcherEmitter, but it also takes making local changes manually teaching it about which of our predicates are mutually exclusive. I do hope to be able to make it catching that sort of problems without manual help at some point later (as part of https://github.com/llvm/llvm-project/issues/69256), but it currently feels like a long way to go.

**Comment by kosarev** [comment]
Fixed pattern predicates as well to pass codegen tests (must have missed the failures). Please take another look.

File: llvm/lib/Target/AMDGPU/VOP3PInstructions.td [hunk 1]
```diff
@@ -360,7 +360,7 @@ class UDot2Pat<Instruction Inst> : GCNPat <
                              (and i32:$src1, (i32 65535)))
    ),
   (Inst (i32 8), $src0, (i32 8), $src1, (i32 8), $src2, (i1 0))> {
-  let SubtargetPredicate = !cast<VOP_Pseudo>(Inst).SubtargetPredicate;
+  let Predicates = !cast<VOP_Pseudo>(Inst).Predicates;
```
Comment (jayfoad) [inline llvm/lib/Target/AMDGPU/VOP3PInstructions.td:363]: As a follow up maybe change the template argument to `VOP_Pseudo Inst` to avoid the cast?

**Review by jayfoad — APPROVED** [review]
LGTM.

File: llvm/lib/Target/AMDGPU/VOP3PInstructions.td (hunk 1 above)
Comment (kosarev) [reply llvm/lib/Target/AMDGPU/VOP3PInstructions.td:363]: https://github.com/llvm/llvm-project/pull/71802

**Pull request merged**